// @vue/language-core@2.0.6/lib/generators/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mergeFeatureSettings = exports.enableAllFeatures = exports.disableAllFeatures = exports.getStack = exports.withStack = void 0;
function withStack(code) {
    return [code, getStack()];
}
exports.withStack = withStack;
// TODO: import from muggle-string
function getStack() {
    const stack = new Error().stack;
    let source = stack.split('\n')[3].trim();
    if (source.endsWith(')')) {
        source = source.slice(source.lastIndexOf('(') + 1, -1);
    }
    else {
        source = source.slice(source.lastIndexOf(' ') + 1);
    }
    return source;
}
exports.getStack = getStack;
function disableAllFeatures(override) {
    return {
        verification: false,
        completion: false,
        semantic: false,
        navigation: false,
        structure: false,
        format: false,
        ...override,
    };
}
exports.disableAllFeatures = disableAllFeatures;
function enableAllFeatures(override) {
    return {
        verification: true,
        completion: true,
        semantic: true,
        navigation: true,
        structure: true,
        format: true,
        ...override,
    };
}
exports.enableAllFeatures = enableAllFeatures;
function mergeFeatureSettings(base, ...others) {
    const result = { ...base };
    for (const info of others) {
        for (const key in info) {
            const value = info[key];
            if (value) {
                result[key] = value;
            }
        }
    }
    return result;
}
exports.mergeFeatureSettings = mergeFeatureSettings;
