// @vue/language-core@2.0.6/lib/languageModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createVueLanguagePlugin = void 0;
const language_core_1 = require("@volar/language-core");
const plugins_1 = require("./plugins");
const vueFile_1 = require("./virtualFile/vueFile");
const CompilerDOM = require("@vue/compiler-dom");
const CompilerVue2 = require("./utils/vue2TemplateCompiler");
const normalFileRegistries = [];
const holderFileRegistries = [];
function getVueFileRegistry(isGlobalTypesHolder, key, plugins) {
    const fileRegistries = isGlobalTypesHolder ? holderFileRegistries : normalFileRegistries;
    let fileRegistry = fileRegistries.find(r => r.key === key
        && r.plugins.length === plugins.length
        && r.plugins.every(plugin => plugins.includes(plugin)))?.files;
    if (!fileRegistry) {
        fileRegistry = new Map();
        fileRegistries.push({
            key: key,
            plugins: plugins,
            files: fileRegistry,
        });
    }
    return fileRegistry;
}
function getFileRegistryKey(compilerOptions, vueCompilerOptions, plugins) {
    const values = [
        ...Object.keys(vueCompilerOptions)
            .sort()
            .filter(key => key !== 'plugins')
            .map(key => [key, vueCompilerOptions[key]]),
        [...new Set(plugins.map(plugin => plugin.requiredCompilerOptions ?? []).flat())]
            .sort()
            .map(key => [key, compilerOptions[key]]),
    ];
    return JSON.stringify(values);
}
function createVueLanguagePlugin(ts, getFileName, isValidGlobalTypesHolder, compilerOptions, vueCompilerOptions, codegenStack = false) {
    const allowLanguageIds = new Set(['vue']);
    const pluginContext = {
        modules: {
            '@vue/compiler-dom': vueCompilerOptions.target < 3
                ? {
                    ...CompilerDOM,
                    compile: CompilerVue2.compile,
                }
                : CompilerDOM,
            typescript: ts,
        },
        compilerOptions,
        vueCompilerOptions,
        codegenStack,
        globalTypesHolder: undefined,
    };
    const plugins = (0, plugins_1.getDefaultVueLanguagePlugins)(pluginContext);
    if (vueCompilerOptions.extensions.includes('.md')) {
        allowLanguageIds.add('markdown');
    }
    if (vueCompilerOptions.extensions.includes('.html')) {
        allowLanguageIds.add('html');
    }
    return {
        createVirtualCode(fileId, languageId, snapshot) {
            if (allowLanguageIds.has(languageId)) {
                const fileName = getFileName(fileId);
                if (!pluginContext.globalTypesHolder && isValidGlobalTypesHolder(fileName)) {
                    pluginContext.globalTypesHolder = fileName;
                }
                const fileRegistry = getFileRegistry(pluginContext.globalTypesHolder === fileName);
                const code = fileRegistry.get(fileId);
                if (code) {
                    code.update(snapshot);
                    return code;
                }
                else {
                    const code = new vueFile_1.VueGeneratedCode(fileName, languageId, snapshot, vueCompilerOptions, plugins, ts, codegenStack);
                    fileRegistry.set(fileId, code);
                    return code;
                }
            }
        },
        updateVirtualCode(_fileId, code, snapshot) {
            code.update(snapshot);
            return code;
        },
        disposeVirtualCode(fileId, code, files) {
            const isGlobalTypesHolder = code.fileName === pluginContext.globalTypesHolder;
            const fileRegistry = getFileRegistry(isGlobalTypesHolder);
            fileRegistry.delete(fileId);
            if (isGlobalTypesHolder) {
                pluginContext.globalTypesHolder = undefined;
                const fileRegistry2 = getFileRegistry(false);
                for (const [fileId, code] of fileRegistry2) {
                    if (isValidGlobalTypesHolder(code.fileName)) {
                        pluginContext.globalTypesHolder = code.fileName;
                        fileRegistry2.delete(fileId);
                        // force dirty
                        files?.delete(fileId);
                        files?.set(fileId, code.languageId, code.snapshot);
                        break;
                    }
                }
            }
        },
        typescript: {
            extraFileExtensions: vueCompilerOptions.extensions.map(ext => ({
                extension: ext.slice(1),
                isMixedContent: true,
                scriptKind: 7,
            })),
            getScript(rootVirtualCode) {
                for (const code of (0, language_core_1.forEachEmbeddedCode)(rootVirtualCode)) {
                    if (code.id.startsWith('script_')) {
                        const lang = code.id.substring('script_'.length);
                        return {
                            code,
                            extension: '.' + lang,
                            scriptKind: lang === 'js' ? ts.ScriptKind.JS
                                : lang === 'jsx' ? ts.ScriptKind.JSX
                                    : lang === 'tsx' ? ts.ScriptKind.TSX
                                        : ts.ScriptKind.TS,
                        };
                    }
                }
            },
        },
    };
    function getFileRegistry(isGlobalTypesHolder) {
        return getVueFileRegistry(isGlobalTypesHolder, getFileRegistryKey(compilerOptions, vueCompilerOptions, plugins), vueCompilerOptions.plugins);
    }
}
exports.createVueLanguagePlugin = createVueLanguagePlugin;
